In [0]:
import pandas as pd
import numpy as np
import sklearn
from sklearn.decomposition import TruncatedSVD

## prepareing the Data

In [0]:
columns = ['user_id', 'item_id', 'rating', 'timestamp']
frame = pd.read_csv('/content/drive/My Drive/linkdin-notebook/movielens/u.data', sep = '\t', names = columns)

In [3]:
frame.head()

user_id  item_id  rating  timestamp
0      196      242       3  881250949
1      186      302       3  891717742
2       22      377       1  878887116
3      244       51       2  880606923
4      166      346       1  886397596

In [5]:
columns = ['item_id', 'movie_title', 'release_date', "Video-release-date","IMDb URL",'unknown','Action', 'Adventure', 'Animation','childrens', 'Comedy','crime', 'Documentry',"Drama", "Fantasy", "Film-Noir", 'Horror', 'Mucical', "Mystery", "Romance", "Sci-Fi", "Thriller", "War", "Western"]
movies = pd.read_csv("/content/drive/My Drive/linkdin-notebook/movielens/u.item",sep = '|', names = columns, encoding ='latin-1')
movies.head()

item_id        movie_title release_date  ...  Thriller War  Western
0        1   Toy Story (1995)  01-Jan-1995  ...         0   0        0
1        2   GoldenEye (1995)  01-Jan-1995  ...         1   0        0
2        3  Four Rooms (1995)  01-Jan-1995  ...         1   0        0
3        4  Get Shorty (1995)  01-Jan-1995  ...         0   0        0
4        5     Copycat (1995)  01-Jan-1995  ...         1   0        0

[5 rows x 24 columns]

In [6]:
movie_names = movies[['item_id', 'movie_title']]
movie_names.head()

item_id        movie_title
0        1   Toy Story (1995)
1        2   GoldenEye (1995)
2        3  Four Rooms (1995)
3        4  Get Shorty (1995)
4        5     Copycat (1995)

In [8]:
combined_movie_data = pd.merge(frame, movie_names, on='item_id')
combined_movie_data.head()

user_id  item_id  rating  timestamp   movie_title
0      196      242       3  881250949  Kolya (1996)
1       63      242       3  875747190  Kolya (1996)
2      226      242       5  883888671  Kolya (1996)
3      154      242       3  879138235  Kolya (1996)
4      306      242       5  876503793  Kolya (1996)

## Sorting bases on popularity of how many times a movie is rated

In [9]:
combined_movie_data.groupby('item_id')['rating'].count().sort_values(ascending = False)

item_id
50      583
258     509
100     508
181     507
294     485
       ... 
1452      1
1593      1
1447      1
814       1
1682      1
Name: rating, Length: 1682, dtype: int64

In [10]:
## here the most rated movie which is rated 583 times with item_id is 50 is star wars
filter = combined_movie_data['item_id']==50
combined_movie_data[filter]['movie_title'].unique()

array(['Star Wars (1977)'], dtype=object)

## Building the utility matrix

In [11]:
rating_crosstab = combined_movie_data.pivot_table(values = 'rating', index = 'user_id', columns = 'movie_title', fill_value=0)
rating_crosstab.head()

movie_title  'Til There Was You (1997)  ...  Á köldum klaka (Cold Fever) (1994)
user_id                                 ...                                    
1                                    0  ...                                   0
2                                    0  ...                                   0
3                                    0  ...                                   0
4                                    0  ...                                   0
5                                    0  ...                                   0

[5 rows x 1664 columns]

In [12]:
star_wars_rating = rating_crosstab['Star Wars (1977)']
star_wars_rating

user_id
1      5
2      5
3      0
4      5
5      4
      ..
939    0
940    4
941    0
942    5
943    4
Name: Star Wars (1977), Length: 943, dtype: int64

## Transposing the matrix

In [13]:
rating_crosstab.shape

(943, 1664)

In [14]:
X = rating_crosstab.values.T
X.shape



(1664, 943)

## Decomposing the Matrix


IN this first we Transposed the metrics before to set movie namees to row and user id to columns now, once this is done we decompose the 943 user rating to 12 using TruncatedSVD 

In [15]:
## random_state to get every time same result every time we run the shell
SVD = TruncatedSVD(n_components=12, random_state=17)
resultant_matrix = SVD.fit_transform(X)
resultant_matrix.shape

(1664, 12)

## Generating Correlation Matrix

we generate the correlation matrix of how the each movie is related to each other using pearson correlation in which we have made the corr_matrix of shape 1664, 1664 as this it shows how each movie in list of 1664 movies is related to each other

for ex how much star wars related to each of 1664 movies we extract that column and see the names of mostly correlated movies

In [16]:
corr_mat = np.corrcoef(resultant_matrix)
corr_mat.shape

(1664, 1664)

## Isolating the stars Wars from the correlation matrix

In [17]:
movie_names = rating_crosstab.columns
movies_list = list(movie_names)
star_wars = movies_list.index('Star Wars (1977)')
star_wars

1398

In [18]:
corr_star_wars = corr_mat[star_wars]
corr_star_wars.shape

(1664,)

## Recommending a highly correlated movies

In [19]:
list(movie_names[(corr_star_wars<1.0) & (corr_star_wars>0.9)])

['Die Hard (1988)',
 'Empire Strikes Back, The (1980)',
 'Fugitive, The (1993)',
 'Raiders of the Lost Ark (1981)',
 'Return of the Jedi (1983)',
 'Star Wars (1977)',
 'Terminator 2: Judgment Day (1991)',
 'Terminator, The (1984)',
 'Toy Story (1995)']

In [20]:
list(movie_names[(corr_star_wars<1.0) & (corr_star_wars>0.95)])

['Return of the Jedi (1983)', 'Star Wars (1977)']